Задание 5, часть 2 (Доверительный интервал на основе ЦПТ)

In [1]:
import pandas as pd
import numpy as np
from scipy.stats import norm
from statsmodels.stats.weightstats import _zconfint_generic

#Загрузка данных
df = pd.read_csv('C:\\Users\\Roadmarshal\\Downloads\\telecom_churn.csv')

#Разделение данных
churned = df[df['churn'] == True]['total day calls']
not_churned = df[df['churn'] == False]['total day calls']

#Расчет среднего и стандартной ошибки
# Cреднее и стандартная ошибка для группы "ушли"
mean_churned = np.mean(churned)
stderr_churned = np.std(churned, ddof=1) / np.sqrt(len(churned))

# Cреднее и стандартная ошибка для группы "не ушли"
mean_not_churned = np.mean(not_churned)
stderr_not_churned = np.std(not_churned, ddof=1) / np.sqrt(len(not_churned))

#Строим интервалы, используя ЦПТ
# Для группы "ушли"
ci_lower_churned = mean_churned - 1.96 * stderr_churned
ci_upper_churned = mean_churned + 1.96 * stderr_churned

# Для группы "не ушли"
ci_lower_not_churned = mean_not_churned - 1.96 * stderr_not_churned
ci_upper_not_churned = mean_not_churned + 1.96 * stderr_not_churned

#Использование statsmodels для расчета интервалов
# Для группы "ушли"
ci_churned_statsmodels = _zconfint_generic(mean_churned, stderr_churned, 0.05, 'two-sided')

# Для группы "не ушли"
ci_not_churned_statsmodels = _zconfint_generic(mean_not_churned, stderr_not_churned, 0.05, 'two-sided')

#Сравнение интервалов
print("Доверительные интервалы для среднего значения total day calls - 'ушли' (ЦПТ): [", format(ci_lower_churned, ".2f"), ",", format(ci_upper_churned, ".2f"), "]")
print("Доверительные интервалы для среднего значения total day calls - 'ушли' (statsmodels):", tuple(map(lambda x: format(x, ".2f"), ci_churned_statsmodels)))
print()
print("Доверительные интервалы для среднего значения total day calls - 'не ушли' (ЦПТ): [", format(ci_lower_not_churned, ".2f"), ",", format(ci_upper_not_churned, ".2f"), "]")
print("Доверительные интервалы для среднего значения total day calls - 'не ушли' (statsmodels):", tuple(map(lambda x: format(x, ".2f"), ci_not_churned_statsmodels)))

Доверительные интервалы для среднего значения total day calls - 'ушли' (ЦПТ): [ 99.41 , 103.26 ]
Доверительные интервалы для среднего значения total day calls - 'ушли' (statsmodels): ('99.41', '103.26')

Доверительные интервалы для среднего значения total day calls - 'не ушли' (ЦПТ): [ 99.56 , 101.01 ]
Доверительные интервалы для среднего значения total day calls - 'не ушли' (statsmodels): ('99.56', '101.01')


1) Для группы "ушли":
99.41 , 103.26 (ЦПТ) и 99.41, 103.26 (statsmodels): Эти интервалы говорят нам, что с 95% вероятностью среднее количество звонков в день для ушедших клиентов лежит между 99.41 и 103.26 звонков. Это означает, что если бы мы многократно выбирали случайные выборки ушедших клиентов и для каждой выборки рассчитывали среднее количество дневных звонков, то примерно в 95% случаев это среднее попадало бы в упомянутый интервал.
2) Для группы "не ушли":
99.56 , 101.01 (ЦПТ) и 99.56, 101.01 (statsmodels): Аналогично, эти доверительные интервалы показывают, что с 95% вероятностью среднее количество звонков в день для клиентов, которые остались с компанией, будет в пределах от 99.56 до 101.01 звонков. Это означает, что если бы мы повторяли процесс измерения для множества различных выборок оставшихся клиентов, то приблизительно в 95% этих измерений среднее значение попадало бы в данный интервал.

Видно, что доверительные интервалы для обеих групп перекрываются, что может указывать на то, что значимых различий в среднем количестве дневных звонков между ушедшими и не ушедшими клиентами нет или они незначительны